### PR TITLE
feat(eventing): eventing integration and pool events

### PR DESCRIPTION
**Pull request by datacore-vvarakantham**
Eventing integration and pool create, delete events generation.

### CODE REVIEW

File: io-engine/src/logger.rs [hunk 1]
```diff
@@ -420,26 +425,45 @@ impl LogFormat {
...
     // Create a default subscriber.
-    let builder = tracing_subscriber::fmt::Subscriber::builder()
+    let builder = tracing_subscriber::fmt::layer()
         .with_span_events(FmtSpan::FULL)
         .event_format(format);
 
-    let subscriber = match EnvFilter::try_from_default_env() {
-        Ok(filter) => builder.with_env_filter(filter).finish(),
-        Err(_) => builder.with_env_filter(level).finish(),
+    let filter = match EnvFilter::try_from_default_env() {
```
Comment (tiagolobocastro) [inline io-engine/src/logger.rs:448]: Should we filter out "mbus-events-target" from stdout?

File: protoc-3.15.8-linux-x86_64.zip
Comment (tiagolobocastro) [inline protoc-3.15.8-linux-x86_64.zip:1]: this is not needed I'm assuming

**Review by chriswldenyer — APPROVED** [review]
Looks believable, but it looks like you need to remove the zip file.

File: protoc-3.15.8-linux-x86_64.zip
Comment (datacore-vvarakantham) [reply protoc-3.15.8-linux-x86_64.zip:1]: Yes, it is not needed. I will remove it.

File: io-engine/src/logger.rs [hunk 2]
```diff
@@ -420,26 +425,45 @@ impl LogFormat {
...
+        Ok(filter) => filter,
+        Err(_) => tracing_subscriber::EnvFilter::new(level),
+    };
+
+    // Get the optional eventing layer.
+    let events_layer = match events_url {
+        Some(url) => {
+            let events_filter =
+                Targets::new().with_target("mbus-events-target", Level::INFO);
```
Comment (sinhaashish) [inline io-engine/src/logger.rs]: Better use constants for `mbus-events-target` and `io-engine`

File: io-engine/src/logger.rs (hunk 1 above)
Comment (datacore-vvarakantham) [reply io-engine/src/logger.rs:448]: Implemented excluding eventing traces now.

File: io-engine/src/lib.rs [hunk 3]
```diff
@@ -45,3 +47,9 @@ pub extern "C" fn cps_init() {
     bdev::nexus::register_module();
     bdev::null_ng::register();
 }
+
+/// Event trait definition for creating events.
+pub(crate) trait Event {
+    /// Create event message.
+    fn event(&self, event_action: Option<EventAction>) -> EventMessage;
```
Comment (tiagolobocastro) [inline io-engine/src/lib.rs]: Why is action optional?

**Review by tiagolobocastro — APPROVED** [review]

File: io-engine/src/lib.rs (hunk 3 above)
Comment (datacore-vvarakantham) [reply io-engine/src/lib.rs]: I've allowed for event_action to be optional in case it isn't required for any particular event generation. But there are no events with event_action of none for now. I can remove the optional if necessary.

Comment (tiagolobocastro) [reply io-engine/src/lib.rs]: I see, it's fine then I was just curious.

File: io-engine/src/lvs/lvs_store.rs [hunk 4]
```diff
@@ -848,3 +863,20 @@ impl PtplFileOps for LvsPtpl {
         std::path::PathBuf::from("pool/").join(self.uuid())
     }
 }
+
+// Pool event messages from Lvs data.
+impl Event for Lvs {
+    fn event(&self, event_action: Option<EventAction>) -> EventMessage {
+        let event_meta = EventMeta::new();
```
Comment (dsavitskiy) [inline io-engine/src/lvs/lvs_store.rs]: You should implement Default for EventMeta, if you have new() without args.

File: io-engine/src/lvs/lvs_store.rs [hunk 5]
```diff
@@ -848,3 +863,20 @@ impl PtplFileOps for LvsPtpl {
...
+        let event_source = EventSource::new(node_name(&None));
+        EventMessage {
+            category: EventCategory::Pool as i32,
+            action: event_action.unwrap_or_default() as i32,
+            target: self.name().to_string(),
+            metadata: Some(EventMeta {
+                source: Some(event_source),
+                ..event_meta
```
Comment (dsavitskiy) [inline io-engine/src/lvs/lvs_store.rs]: If you implement Default for EventMeta, you can use `..Default::default()` instead of having a tmp variable "event_meta", making your code cleaner.

Comment (dsavitskiy) [reply io-engine/src/lvs/lvs_store.rs]: Another suggestion, you can add a method like "from_source()" to EventMeta, if creating from source is a frequent operation for meta.

File: io-engine/src/lvs/lvs_store.rs [hunk 6]
```diff
@@ -52,6 +52,16 @@ use crate::{
     pool_backend::PoolArgs,
 };
 
+use events_api::event::{
```
Comment (dsavitskiy) [inline io-engine/src/lvs/lvs_store.rs]: Probably, we should consider moving mbus-specific code (dependent on generated code) to a separate module, like we do for gRPC-specific?

File: io-engine/src/logger.rs [hunk 7]
```diff
@@ -420,26 +428,50 @@ impl LogFormat {
...
+
     LogTracer::init().expect("failed to initialise LogTracer");
 
     // Create a default subscriber.
-    let builder = tracing_subscriber::fmt::Subscriber::builder()
+    let builder = tracing_subscriber::fmt::layer()
         .with_span_events(FmtSpan::FULL)
-        .event_format(format);
+        .event_format(format)
+        .with_filter(filter_fn(|metadata| {
```
Comment (dsavitskiy) [inline io-engine/src/logger.rs:442]: Instead of modifying logger, can we consider using existing functionality to selectively rise and lower level for specific target? Probably, changing RUST_LOG defaults to exclude "mbus-events-target" would work.

If changing RUST_LOG defaults proves to be no good for this particular case, could you move logger changes to separate commit?

**Review by dsavitskiy — CHANGES_REQUESTED** [review]

File: io-engine/src/lvs/lvs_store.rs (hunk 6 above)
Comment (datacore-vvarakantham) [reply io-engine/src/lvs/lvs_store.rs]: Done

File: io-engine/src/lvs/lvs_store.rs (hunk 5 above)
Comment (datacore-vvarakantham) [reply io-engine/src/lvs/lvs_store.rs]: Added from_source() now in dependencies. Will pull in the changes here when the changes are merged in dependencies.

File: io-engine/src/logger.rs (hunk 7 above)
Comment (datacore-vvarakantham) [reply io-engine/src/logger.rs:442]: Changing the defaults to exclude "mbus-event-target" is removing the eventing traces from all the layers and the events are not generated.

Added eventing integration changes to separate commit now.

File: io-engine/src/lvs/lvs_store.rs (hunk 4 above)
Comment (datacore-vvarakantham) [reply io-engine/src/lvs/lvs_store.rs]: I added from_source as suggested in dependencies which also removes the need for tmp variable.

**Review by dsavitskiy — APPROVED** [review]

**Comment by datacore-vvarakantham** [comment]
bors merge
